### PR TITLE
HFE - Fix and refactor RDB load and align RESTORE 

### PR DESCRIPTION
**Pull request by moticless**
* Modify to load `hfield` and not `sds` from the start and let it work directly on `dict`.
* Removed any DB logic from rdbLoadObject() and let it handle the object alone. The caller to rdbLoadObject() is the one to register if needed to DB. Discard the optimized approach of having distinct handling between HT and listpackEx. Both are RDB loading blindly expired fields. We could have discard expired field while RDB loading like we do with expired keys, but then we will need to keep track of which fields expired and later, if entire object loaded successfully, notify the replicas which one to discard. Instead we simplified it by loading everything into a hash and let active expiration do its job. This way we don't need to keep any special replica logic and without bookkeeping in rdbLoadObject() and have same behaviour to listpackex and HT.
* Following previous bullet, remove `rdb_last_load_hash_fields_expired` statistic.
* Align RESTORE usage of `rdbLoadObject()` accordingly.

### CODE REVIEW

File: src/rdb.c [hunk 1]
```diff
@@ -2278,8 +2275,13 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, redisDb* db, int rdbflags,
                 return NULL;
             }
 
-            /* read the field name */
-            if ((field = rdbGenericLoadStringObject(rdb, RDB_LOAD_SDS, &fieldLen)) == NULL) {
+            /* if needed create field with TTL metadata  */
+            if (expireAt !=0)
```
Comment (tezc) [inline src/rdb.c:2287]:
```suggestion
            if (expireAt != 0)
```

File: src/rdb.c [hunk 2]
```diff
@@ -2291,111 +2293,78 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, redisDb* db, int rdbflags,
...
+
+                /* Attach expiry to the hash field and register in hash private HFE DS */
+                if ((ret != DICT_ERR) && expireAt) {
+                    dictExpireMetadata *m = (dictExpireMetadata *) dictMetadata(d);
+                    ret = ebAdd(&m->hfe, &hashFieldExpireBucketsType, field, expireAt);
+                }
+
+                if (ret == DICT_ERR) {
+                    rdbReportCorruptRDB("Duplicate hash fields detected");
+                    sdsfree(value);
```
Comment (tezc) [inline src/rdb.c:2371]: don't you need to free dupSearchDict here? or somewhere else before reaching to this point?

File: src/rdb.c [hunk 3]
```diff
@@ -2291,111 +2293,78 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, redisDb* db, int rdbflags,
...
                         }
                     }
 
                     /* don't add the values to the new hash: the next if will catch and the values will be added there */
                 } else {
-                    listpackExAddNew(o, field, value, expireAt);
-                    sdsfree(field);
+                    listpackExAddNew(o, field, hfieldlen(field),
+                                     value, sdslen(value), expireAt);
+                    hfieldFree(field);
```
Comment (tezc) [inline src/rdb.c:2351]: So we used sds functions for hfield before and sanitizer did not warn us? Maybe we need some test cases that would fail on such a mistake.

Comment (sundb) [reply src/rdb.c:2351]: @tezc ASAN reported, but not here.
i think sds and hfield are too similar that in some cases they are the same.

Comment (tezc) [reply src/rdb.c:2351]: Sure I know. Still we can try to add a test where they are not similar. Any mismatch can cause a sanitizer error/test failure. e.g. we can use longer strings

**Comment by tezc** [comment]
Overall, LGTM!
 I just remember Oran was eager to handle expired keys as much as possible while loading.

File: src/rdb.c (hunk 2 above)
Comment (moticless) [reply src/rdb.c:2371]: There is an implicit assumption that once the object is `OBJ_ENCODING_HT`, then `dupSearchDict` got released.

File: src/rdb.c (hunk 1 above)
Comment (moticless) [reply src/rdb.c:2287]: i will modify it if i will have another important change.

**Comment by moticless** [comment]
@tezc , ok. I talked with Oran and looks like we are aligned.

File: src/rdb.c (hunk 2 above)
Comment (tezc) [reply src/rdb.c:2371]: Yeah but looks like it is not the case. It is hard to follow this function but on line 2334, it converts to HT but does not free dupSearchDict.
